Move MakeFile/Dir from HostUtil to host_path vol

codenrhoden opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind cleanup


**What this PR does / why we need it**:
This clarification was requested as part of https://github.com/kubernetes/utils/pull/100

The MakeFile and MakeDir methods in the HostUtil interface only had one
caller -- the Host Path volume plugin. This patch relocates MakeFile and
MakeDir to the Host Path plugin itself.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

codenrhoden commented:
/assign @msau42 

Michelle, I think it's worth considering if the documented behavior is the behavior we *want*. As it is , if you create a directory, and the given path is already there and happens to be a file instead of a directory, I think it will succeed (because this will just be an `IsExist` error. I'll write a quick test program to verify if that's true.

codenrhoden commented:
Okay, so I wrote a quick Go program to test this behavior...

For `MakeDir`, if you are given a path and it already exists as a file, it does produce an error:
`mkdir /tmp/travis: not a directory`

However, for `MakeFile`, if you are given a path and it already exists as a file, it does not produce an error at all. `os.OpenFile` will happily open a directory and give you a file handle back.

Furthermore, both calls pass a certain set of permissions. Nothing is done to ensure that if the path already exists, that those permissions are set correctly. :/

Should we try to fix this code correctly? Namely...

Make each function take in an optional value to set file permissions (which are only honored on *nix systems, i believe), and before creating the pathname,

- check if it exists first
- check that it is the correct type
- check that is has the correct permissions

Thoughts?

hogepodge commented on pkg/util/mount/mount.go:
```diff
@@ -88,8 +88,10 @@ type HostUtils interface {
 	// GetFileType checks for file/directory/socket/block/character devices.
 	GetFileType(pathname string) (FileType, error)
 	// MakeFile creates an empty file.
+	// If pathname already exists, whether a file or directory, no error is returned.
```
This isn't a comment that needs to be addressed in this patch, but is about the methods in general. If we are making a file or a directory, and a directory or file is there respectively, this would violate the principle of least surprise. I'm glad this patch adds clarification to the documentation to reduce the surprise potential.

hogepodge commented:
/lgtm

codenrhoden commented:
@msau42 Now that [nsenter is going away](https://github.com/kubernetes/kubernetes/pull/81843), I think it's worth revisiting this.

As far as I can tell, the only callers of `MakeFile` and `MakeDir` are within the `hostpath` volume plugin. I don't think that warrants being in this interface, and therefore hostpath should do what it needs directly.

Alternatively, if we do want to keep this, I still have the open questions from https://github.com/kubernetes/kubernetes/pull/81739#issuecomment-523623167. I find the current behavior to be inconsistent, and @hogepodge's comment points out that it just doesn't do what you would expect... I obviously don't want to unexpectedly change any behavior of `hostpath`, so I'd probably just move the functions over to `hostpath`, but I think it's worth discussing what it *should* look like.

msau42 commented:
Agree, let's internalize this to hostpath, and open up a separate issue to track the confusing behavior

codenrhoden committed:
Move MakeFile/Dir from HostUtil to host_path vol

The MakeFile and MakeDir methods in the HostUtil interface only had one
caller -- the Host Path volume plugin. This patch relocates MakeFile and
MakeDir to the Host Path plugin itself.

saad-ali reviewed (approved):
/lgtm
/approve
